akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessImpl.scala

```diff
@@ -0,0 +1,171 @@
+            case Tick =>
+              triggerStartAll()
+              context.log.debug("Periodic ping sent to [{}] processes", identities.size)
+              Behaviors.same
+          }
+        }
+      }
+  }
+
+  final class MessageExtractor[T] extends ShardingMessageExtractor[ShardingEnvelope[T], T] {
+    def entityId(message: ShardingEnvelope[T]): String = message match {
+      case ShardingEnvelope(id, _) => id
+    }
+
+    def shardId(entityId: String): String = entityId
+
+    def unwrapMessage(message: ShardingEnvelope[T]): T = message.message
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class ShardedDaemonProcessImpl(system: ActorSystem[_])
+    extends javadsl.ShardedDaemonProcess
+    with scaladsl.ShardedDaemonProcess {
+
+  import ShardedDaemonProcessImpl._
+
+  def init[T](name: String, numberOfInstances: Int, behaviorFactory: Int => Behavior[T])(
+      implicit classTag: ClassTag[T]): Unit = {
+    init(name, numberOfInstances, behaviorFactory, ShardedDaemonProcessSettings(system), None)(classTag)
+  }
+
+  def init[T](
+      name: String,
+      numberOfInstances: Int,
+      behaviorFactory: Int => Behavior[T],
+      settings: ShardedDaemonProcessSettings,
+      stopMessage: Option[T])(implicit classTag: ClassTag[T]): Unit = {
+
+    val entityTypeKey = EntityTypeKey[T](s"sharded-daemon-process-$name")
+
+    // One shard per actor identified by the numeric id encoded in the entity id
+    val numberOfShards = numberOfInstances
+    val entityIds = (0 to numberOfInstances).map(_.toString)
+
+    val shardingSettings = {
+      val shardingBaseSettings =
+        settings.shardingSettings match {
+          case None =>
+            // defaults in akka.cluster.sharding but allow overrides specifically for actor-set
+            ClusterShardingSettings.fromConfig(
+              system.settings.config.getConfig("akka.cluster.sharded-daemon-process.sharding"))
+          case Some(shardingSettings) => shardingSettings
+        }
+
+      new ClusterShardingSettings(
+        numberOfShards,
+        shardingBaseSettings.role,
+        shardingBaseSettings.dataCenter,
+        false, // remember entities disabled
+        "",
+        "",
+        Duration.Zero, // passivation disabled
+        shardingBaseSettings.shardRegionQueryTimeout,
+        StateStoreModeDData,
+        shardingBaseSettings.tuningParameters,
+        shardingBaseSettings.coordinatorSingletonSettings)
+    }
+
+    val entity = Entity(entityTypeKey)(ctx => behaviorFactory(ctx.entityId.toInt))
+      .withSettings(shardingSettings)
+      .withMessageExtractor(new MessageExtractor)
+
+    val entityWithStop = stopMessage match {
+      case Some(stop) => entity.withStopMessage(stop)
+      case None       => entity
+    }
+
+    val shardingRef = ClusterSharding(system).init(entityWithStop)
+
+    val nodeRoles = Cluster(system).selfMember.roles
+    if (shardingSettings.role.forall(nodeRoles)) {
```

Comment:
shouldn't all of it, including the ClusterSharding.init be conditioned by the role, otherwise it will start a region proxy on nodes where it's not needed?

Reply:
but that depends on the question above regarding the region ActorRef, if we return the ActorRef we need to init to return proxy refs

Reply:
That's true, if we agree on keeping it a `Unit` startup method all setup should go in this if.

Reply:
Updated to avoid setting up if not the configured role